### PR TITLE
ADBDEV-3098-10: ORCA produces bogus plan for queries with CTE during handling distribution for Sequence children

### PR DESCRIPTION
**Pull request by RekGRpth**
With some conditions ORCA generates `Redistribute Motion` from all segments to one segment which can not executed because inconsistency CTE producers and consumers.
For example, create tables
```sql
CREATE TABLE d (a int, b int) DISTRIBUTED BY (a);
CREATE TABLE r (a int, b varchar(255)) DISTRIBUTED REPLICATED;
```
and run query
```sql
EXPLAIN (ANALYZE off, COSTS off, VERBOSE off)
WITH e AS (
    SELECT DISTINCT b FROM d
), h AS (
    SELECT a FROM d JOIN e f USING (b) JOIN e USING (b)
) SELECT b FROM h JOIN r USING (a);
```
that gives bogus plan
```sql
                                        QUERY PLAN                                        
------------------------------------------------------------------------------------------
 Gather Motion 1:1  (slice4; segments: 1)
   ->  Sequence
         ->  Shared Scan (share slice:id 4:0)
               ->  Materialize
                     ->  GroupAggregate
                           Group Key: d_1.b
                           ->  Sort
                                 Sort Key: d_1.b
                                 ->  Redistribute Motion 3:1  (slice3; segments: 3)
                                       Hash Key: d_1.b
                                       ->  Seq Scan on d d_1
         ->  Hash Join
               Hash Cond: (r.a = d.a)
               ->  Seq Scan on r
               ->  Hash
                     ->  Broadcast Motion 3:1  (slice2; segments: 3)
                           ->  Hash Join
                                 Hash Cond: (d.b = share0_ref2.b)
                                 ->  Hash Join
                                       Hash Cond: (d.b = share0_ref3.b)
                                       ->  Redistribute Motion 3:3  (slice1; segments: 3)
                                             Hash Key: d.b
                                             ->  Seq Scan on d
                                       ->  Hash
                                             ->  Shared Scan (share slice:id 2:0)
                                 ->  Hash
                                       ->  Shared Scan (share slice:id 2:0)
 Optimizer: Pivotal Optimizer (GPORCA)
(28 rows)
```
In this plan `Shared Scan (share slice:id 4:0)` is CTE producer that is executed only on one segment because `Redistribute Motion 3:1  (slice3; segments: 3)`, but CTE consumers (`Shared Scan (share slice:id 2:0)`) are executed on all 3 segments and then query hangs.

We get Redistribute N:1 because following. When translating expression to DXL, ORCA sets one segment for input array if strict or tainted replicated https://github.com/arenadata/gpdb/blob/2104e0653b8a98ab42ce3fbd2324cc1b750a491b/src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp#L8236-L8264 and this one segment is used as output array for below motions including redistribute and broadcast.

Solution is require enforcer for non-singleton distribution (that leads to finalizing current optimization job and creating new optimization job with enforcer) when child delivers replicated distribution and there is CTE. It is gives right plan
```sql
                                        QUERY PLAN                                        
------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice4; segments: 3)
   ->  Sequence
         ->  Shared Scan (share slice:id 4:0)
               ->  Materialize
                     ->  GroupAggregate
                           Group Key: d_1.b
                           ->  Sort
                                 Sort Key: d_1.b
                                 ->  Redistribute Motion 3:3  (slice3; segments: 3)
                                       Hash Key: d_1.b
                                       ->  Seq Scan on d d_1
         ->  Hash Join
               Hash Cond: (r.a = d.a)
               ->  Result
                     ->  Seq Scan on r
               ->  Hash
                     ->  Redistribute Motion 3:3  (slice2; segments: 3)
                           Hash Key: d.a
                           ->  Hash Join
                                 Hash Cond: (d.b = share0_ref2.b)
                                 ->  Hash Join
                                       Hash Cond: (d.b = share0_ref3.b)
                                       ->  Redistribute Motion 3:3  (slice1; segments: 3)
                                             Hash Key: d.b
                                             ->  Seq Scan on d
                                       ->  Hash
                                             ->  Shared Scan (share slice:id 2:0)
                                 ->  Hash
                                       ->  Shared Scan (share slice:id 2:0)
 Optimizer: Pivotal Optimizer (GPORCA)
(30 rows)
```
Upstream master (gpdb 7) also is affected by this problem, but with other error
```sql
FATAL:  Unexpected internal error (assert.c:44)
DETAIL:  FailedAssertion("!(pMotion->numHashSegments < motion_recv)", File: "explain.c", Line: 2513)
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
```

### CODE REVIEW

File: src/backend/gporca/data/dxl/minidump/CTAS_OrderedAgg_multiple_cols.mdp [hunk 1]
```diff
@@ -1808,7 +1808,7 @@ EXPLAIN CREATE TABLE test AS
         </dxl:LogicalProject>
       </dxl:LogicalCTAS>
     </dxl:Query>
-    <dxl:Plan Id="0" SpaceSize="16799099328000">
```
Comment (Stolb27) [inline src/backend/gporca/data/dxl/minidump/CTAS_OrderedAgg_multiple_cols.mdp]: What is SpaceSize? Why the given patch affects this query plan?

File: src/backend/gporca/libgpopt/src/base/CEnfdDistribution.cpp [hunk 2]
```diff
@@ -149,6 +150,21 @@ CEnfdDistribution::Epet(CExpressionHandle &exprhdl, CPhysical *popPhysical,
 			return EpetProhibited;
 		}
 
+		if (CDistributionSpec::EdtNonSingleton == m_pds->Edt() &&
+			CDistributionSpec::EdtStrictReplicated == pds->Edt())
```
Comment (Stolb27) [inline src/backend/gporca/libgpopt/src/base/CEnfdDistribution.cpp]: Can you describe meaning of these distributions for the bogus plan provided in the ticket description?

**Comment by Stolb27** [comment]
1. Please, describe how we get Redistribute N:1 without this patch.
2. Provide formatted commit description
3. I think, it would be nice to provide test case as ORCA mdp file instead of/in addition to regression test.

> that leads to finalizing current optimization job and creating new optimization job with enforcer

Please, explain how ORCA chose enforcer type? Why Redistribute Motion in our case?

File: src/backend/gporca/data/dxl/minidump/CTAS_OrderedAgg_multiple_cols.mdp (hunk 1 above)
Comment (RekGRpth) [reply src/backend/gporca/data/dxl/minidump/CTAS_OrderedAgg_multiple_cols.mdp]: it is count of alternatives https://github.com/arenadata/gpdb/blob/46f40f274b7ebcc5a8d42076d49c786849931d47/src/backend/gporca/libgpopt/src/engine/CEngine.cpp#L694-L695

Comment (Stolb27) [reply src/backend/gporca/data/dxl/minidump/CTAS_OrderedAgg_multiple_cols.mdp]: Why this patch decrease alternatives count? These alternatives were incorrect? Can you proof it?

Comment (RekGRpth) [reply src/backend/gporca/data/dxl/minidump/CTAS_OrderedAgg_multiple_cols.mdp]: > Why this patch decrease alternatives count?

not only descrease: https://github.com/arenadata/gpdb/pull/434/files#diff-9040fc5f4b8fccdcc163c373e9325c1781d11028d016f6ff88213c6b09e5a366L511

Comment (RekGRpth) [reply src/backend/gporca/data/dxl/minidump/CTAS_OrderedAgg_multiple_cols.mdp]: I return extra condition and only 2 mpd are affected now

Comment (InnerLife0) [reply src/backend/gporca/data/dxl/minidump/CTAS_OrderedAgg_multiple_cols.mdp]: That doesn't answer Vasiliy's question at all.

File: src/backend/gporca/libgpopt/src/base/CEnfdDistribution.cpp (hunk 2 above)
Comment (Stolb27) [reply src/backend/gporca/libgpopt/src/base/CEnfdDistribution.cpp]: As far I understand, we require enforcement for any subtree with Strict Replicated distribution, when Non Singleton distribution is requested. But it seems to be incorrect in the common case. How enforcement is chosen?

File: src/backend/gporca/data/dxl/minidump/ReplicatedTableCTE.mdp [hunk 3]
```diff
@@ -1443,7 +1443,7 @@
         </dxl:LogicalJoin>
       </dxl:LogicalCTEAnchor>
     </dxl:Query>
-    <dxl:Plan Id="0" SpaceSize="8">
+    <dxl:Plan Id="0" SpaceSize="7">
```
Comment (RekGRpth) [inline src/backend/gporca/data/dxl/minidump/ReplicatedTableCTE.mdp]: it is interesting that in this case CPhysicalMotionRandom is added in Group 1 but Number of plan alternatives decreases

File: src/backend/gporca/data/dxl/minidump/CTAS_OrderedAgg_multiple_cols.mdp (hunk 1 above)
Comment (RekGRpth) [reply src/backend/gporca/data/dxl/minidump/CTAS_OrderedAgg_multiple_cols.mdp]: > Why the given patch affects this query plan?

given patch does not affect this query plan any more

Comment (RekGRpth) [reply src/backend/gporca/data/dxl/minidump/CTAS_OrderedAgg_multiple_cols.mdp]: > only 2 mpd are affected

changing check from cte consumer to cte producer eliminate these 2 mpds

Comment (RekGRpth) [reply src/backend/gporca/data/dxl/minidump/CTAS_OrderedAgg_multiple_cols.mdp]: is this question resolved?

Comment (InnerLife0) [reply src/backend/gporca/data/dxl/minidump/CTAS_OrderedAgg_multiple_cols.mdp]: For me, I didn't check latest commits to see why mdp files are unchanged again.

**Comment by RekGRpth** [comment]
> 1. Please, describe how we get Redistribute N:1 without this patch.

When translating expression to DXL, ORCA sets one segment for input array if strict or tainted replicated https://github.com/arenadata/gpdb/blob/2104e0653b8a98ab42ce3fbd2324cc1b750a491b/src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp#L8236-L8264 and this one segment is used as output array for below motions including redistribute and broadcast

**Comment by RekGRpth** [comment]
> Why Redistribute Motion in our case?

I think that redistribute is default enforcer and in our case patch does not allow using broadcast enforcer

File: src/backend/gporca/libgpopt/src/base/CEnfdDistribution.cpp (hunk 2 above)
Comment (RekGRpth) [reply src/backend/gporca/libgpopt/src/base/CEnfdDistribution.cpp]: > But it seems to be incorrect in the common case

Yes and I do that only if CTE producer exists in plan

Comment (RekGRpth) [reply src/backend/gporca/libgpopt/src/base/CEnfdDistribution.cpp]: > How enforcement is chosen?

https://github.com/arenadata/gpdb/blob/46f40f274b7ebcc5a8d42076d49c786849931d47/src/backend/gporca/libgpopt/src/search/CJobGroupExpressionOptimization.cpp#L597-L603

Comment (RekGRpth) [reply src/backend/gporca/libgpopt/src/base/CEnfdDistribution.cpp]: > Can you describe meaning of these distributions for the bogus plan provided in the ticket description?

this corresponds to `Seq Scan on r` in our plan

Comment (Stolb27) [reply src/backend/gporca/libgpopt/src/base/CEnfdDistribution.cpp]: > Yes and I do that only if CTE producer exists in plan

So, any Broadcast Motion will be ~~prohibited~~ enforced in queries with CTE (with CTE producer in the first group) regardless of Broadcast motion location in the subtree? I think it's still too broad condition.

> this corresponds to Seq Scan on r in our plan

But the resulting plan doesn't contain any enforced Motions above this scan. Please describe in the **PR description** how this code part works.

**Comment by Stolb27** [comment]
> > Why Redistribute Motion in our case?

> I think that redistribute is default enforcer and in our case patch does not allow using broadcast enforcer

Please, prove it.

> When translating expression to DXL, ORCA sets one segment for input array if strict or tainted replicated 

This must be described in the PR description in text form. Links and code snippets are only useful addition.

**Comment by RekGRpth** [comment]
> I think, it would be nice to provide test case as ORCA mdp file instead of/in addition to regression test.

done

File: src/backend/gporca/libgpopt/src/base/CEnfdDistribution.cpp (hunk 2 above)
Comment (RekGRpth) [reply src/backend/gporca/libgpopt/src/base/CEnfdDistribution.cpp]: > any Broadcast Motion will be

not any, but only under replicated distribution when non-singleton required

**Comment by RekGRpth** [comment]
> Please, prove it.

comparing Physical plan before PR
```sql
+--CPhysicalMotionGather(master)   rows:1   width:335  rebinds:1   cost:2155.005968   origin: [Grp:7, GrpExpr:3]
   +--CPhysicalSequence   rows:1   width:335  rebinds:1   cost:2155.003117   origin: [Grp:7, GrpExpr:2]
      |--CPhysicalCTEProducer (0), Columns: ["b" (1)]   rows:1   width:4  rebinds:1   cost:431.000021   origin: [Grp:28, GrpExpr:1]
      |  +--CPhysicalStreamAgg( Global ) Grp Cols: ["b" (1)], Minimal Grp Cols:["b" (1)], Generates Duplicates :[ 0 ]    rows:1   width:4  rebinds:1   cost:431.000021   origin: [Grp:27, GrpExpr:4]
      |     |--CPhysicalSort  ( (97,1.0), "b" (1), NULLsLast )    rows:1   width:38  rebinds:1   cost:431.000016   origin: [Grp:25, GrpExpr:4]
      |     |  +--CPhysicalMotionHashDistribute HASHED: [ CScalarIdent "b" (1), nulls colocated ], opfamilies: (1977,1.0),   rows:1   width:38  rebinds:1   cost:431.000016   origin: [Grp:25, GrpExpr:2]
      |     |     +--CPhysicalTableScan "d" ("d")   rows:1   width:38  rebinds:1   cost:431.000007   origin: [Grp:25, GrpExpr:1]
      |     +--CScalarProjectList   origin: [Grp:26, GrpExpr:0]
      +--CPhysicalFilter   rows:1   width:335  rebinds:1   cost:1724.002841   origin: [Grp:6, GrpExpr:2]
         |--CPhysicalInnerHashJoin   rows:1   width:335  rebinds:1   cost:1724.002841   origin: [Grp:5, GrpExpr:3]
         |  |--CPhysicalTableScan "r" ("r")   rows:1   width:289  rebinds:1   cost:431.000159   origin: [Grp:1, GrpExpr:1]
         |  |--CPhysicalMotionBroadcast    rows:1   width:46  rebinds:1   cost:1293.000786   origin: [Grp:0, GrpExpr:21]
         |  |  +--CPhysicalInnerHashJoin   rows:1   width:46  rebinds:1   cost:1293.000714   origin: [Grp:0, GrpExpr:17]
         |  |     |--CPhysicalInnerHashJoin   rows:1   width:42  rebinds:1   cost:862.000449   origin: [Grp:18, GrpExpr:3]
         |  |     |  |--CPhysicalMotionHashDistribute HASHED: [ CScalarIdent "b" (37), nulls colocated ], opfamilies: (1977,1.0),   rows:1   width:38  rebinds:1   cost:431.000061   origin: [Grp:8, GrpExpr:3]
         |  |     |  |  +--CPhysicalTableScan "d" ("d")   rows:1   width:38  rebinds:1   cost:431.000021   origin: [Grp:8, GrpExpr:1]
         |  |     |  |--CPhysicalCTEConsumer (0), Columns: ["b" (45)]   rows:1   width:4  rebinds:1   cost:431.000010   origin: [Grp:9, GrpExpr:1]
         |  |     |  +--CScalarCmp (=)   origin: [Grp:13, GrpExpr:0]
         |  |     |     |--CScalarIdent "b" (37)   origin: [Grp:11, GrpExpr:0]
         |  |     |     +--CScalarIdent "b" (45)   origin: [Grp:12, GrpExpr:0]
         |  |     |--CPhysicalCTEConsumer (0), Columns: ["b" (54)]   rows:1   width:4  rebinds:1   cost:431.000010   origin: [Grp:10, GrpExpr:1]
         |  |     +--CScalarCmp (=)   origin: [Grp:15, GrpExpr:0]
         |  |        |--CScalarIdent "b" (37)   origin: [Grp:11, GrpExpr:0]
         |  |        +--CScalarIdent "b" (54)   origin: [Grp:14, GrpExpr:0]
         |  +--CScalarCmp (=)   origin: [Grp:4, GrpExpr:0]
         |     |--CScalarIdent "a" (36)   origin: [Grp:2, GrpExpr:0]
         |     +--CScalarIdent "a" (63)   origin: [Grp:3, GrpExpr:0]
         +--CScalarConst (1)   origin: [Grp:24, GrpExpr:0]
```
and Physical plan after PR
```sql
+--CPhysicalMotionGather(master)   rows:1   width:335  rebinds:1   cost:2155.004048   origin: [Grp:7, GrpExpr:3]
   +--CPhysicalSequence   rows:1   width:335  rebinds:1   cost:2155.003097   origin: [Grp:7, GrpExpr:2]
      |--CPhysicalCTEProducer (0), Columns: ["b" (1)]   rows:1   width:4  rebinds:1   cost:431.000021   origin: [Grp:28, GrpExpr:1]
      |  +--CPhysicalStreamAgg( Global ) Grp Cols: ["b" (1)], Minimal Grp Cols:["b" (1)], Generates Duplicates :[ 0 ]    rows:1   width:4  rebinds:1   cost:431.000021   origin: [Grp:27, GrpExpr:4]
      |     |--CPhysicalSort  ( (97,1.0), "b" (1), NULLsLast )    rows:1   width:38  rebinds:1   cost:431.000016   origin: [Grp:25, GrpExpr:4]
      |     |  +--CPhysicalMotionHashDistribute HASHED: [ CScalarIdent "b" (1), nulls colocated ], opfamilies: (1977,1.0),   rows:1   width:38  rebinds:1   cost:431.000016   origin: [Grp:25, GrpExpr:2]
      |     |     +--CPhysicalTableScan "d" ("d")   rows:1   width:38  rebinds:1   cost:431.000007   origin: [Grp:25, GrpExpr:1]
      |     +--CScalarProjectList   origin: [Grp:26, GrpExpr:0]
      +--CPhysicalFilter   rows:1   width:335  rebinds:1   cost:1724.002991   origin: [Grp:6, GrpExpr:2]
         |--CPhysicalInnerHashJoin   rows:1   width:335  rebinds:1   cost:1724.002991   origin: [Grp:5, GrpExpr:3]
         |  |--CPhysicalMotionHashDistribute HASHED: [ CScalarIdent "a" (63), nulls colocated ], opfamilies: (1977,1.0),   rows:1   width:289  rebinds:1   cost:431.001451   origin: [Grp:1, GrpExpr:3]
         |  |  +--CPhysicalTableScan "r" ("r")   rows:1   width:289  rebinds:1   cost:431.000159   origin: [Grp:1, GrpExpr:1]
         |  |--CPhysicalMotionHashDistribute HASHED: [ CScalarIdent "a" (36), nulls colocated ], opfamilies: (1977,1.0),   rows:1   width:46  rebinds:1   cost:1293.000720   origin: [Grp:0, GrpExpr:22]
         |  |  +--CPhysicalInnerHashJoin   rows:1   width:46  rebinds:1   cost:1293.000714   origin: [Grp:0, GrpExpr:17]
         |  |     |--CPhysicalInnerHashJoin   rows:1   width:42  rebinds:1   cost:862.000449   origin: [Grp:18, GrpExpr:3]
         |  |     |  |--CPhysicalMotionHashDistribute HASHED: [ CScalarIdent "b" (37), nulls colocated ], opfamilies: (1977,1.0),   rows:1   width:38  rebinds:1   cost:431.000061   origin: [Grp:8, GrpExpr:3]
         |  |     |  |  +--CPhysicalTableScan "d" ("d")   rows:1   width:38  rebinds:1   cost:431.000021   origin: [Grp:8, GrpExpr:1]
         |  |     |  |--CPhysicalCTEConsumer (0), Columns: ["b" (45)]   rows:1   width:4  rebinds:1   cost:431.000010   origin: [Grp:9, GrpExpr:1]
         |  |     |  +--CScalarCmp (=)   origin: [Grp:13, GrpExpr:0]
         |  |     |     |--CScalarIdent "b" (37)   origin: [Grp:11, GrpExpr:0]
         |  |     |     +--CScalarIdent "b" (45)   origin: [Grp:12, GrpExpr:0]
         |  |     |--CPhysicalCTEConsumer (0), Columns: ["b" (54)]   rows:1   width:4  rebinds:1   cost:431.000010   origin: [Grp:10, GrpExpr:1]
         |  |     +--CScalarCmp (=)   origin: [Grp:15, GrpExpr:0]
         |  |        |--CScalarIdent "b" (37)   origin: [Grp:11, GrpExpr:0]
         |  |        +--CScalarIdent "b" (54)   origin: [Grp:14, GrpExpr:0]
         |  +--CScalarCmp (=)   origin: [Grp:4, GrpExpr:0]
         |     |--CScalarIdent "a" (36)   origin: [Grp:2, GrpExpr:0]
         |     +--CScalarIdent "a" (63)   origin: [Grp:3, GrpExpr:0]
         +--CScalarConst (1)   origin: [Grp:24, GrpExpr:0]
```
shows that CPhysicalMotionBroadcast from group 0 expression 21 changed to CPhysicalMotionHashDistribute from group 0 expression 22 while group 0 is same before and after PR
```sql
Group 0 (#GExprs: 23):
  0: CLogicalCTEConsumer (1), Columns: ["a" (36)] [ ]
  1: CLogicalNAryJoin [ 8 9 10 17 ]
  2: CLogicalInnerJoin [ 18 10 15 ]
  3: CLogicalInnerJoin [ 19 9 20 ]
  4: CLogicalInnerJoin [ 8 21 22 ]
  5: CLogicalInnerJoin [ 21 8 22 ]
  6: CLogicalInnerJoin [ 9 19 20 ]
  7: CLogicalInnerJoin [ 10 18 23 ]
  8: CLogicalInnerJoin [ 18 10 23 ]
  9: CLogicalInnerJoin [ 10 18 15 ]
  10: CPhysicalInnerHashJoin (High) [ 10 18 15 ]
  11: CPhysicalInnerHashJoin (High) [ 18 10 23 ]
  12: CPhysicalInnerHashJoin (High) [ 10 18 23 ]
  13: CPhysicalInnerHashJoin (High) [ 9 19 20 ]
  14: CPhysicalInnerHashJoin (High) [ 21 8 22 ]
  15: CPhysicalInnerHashJoin (High) [ 8 21 22 ]
  16: CPhysicalInnerHashJoin (High) [ 19 9 20 ]
  17: CPhysicalInnerHashJoin (High) [ 18 10 15 ]
  18: CPhysicalCTEConsumer (1), Columns: ["a" (36)] [ ]
  19: CPhysicalMotionGather(master) [ 0 ]
  20: CPhysicalMotionRandom [ 0 ]
  21: CPhysicalMotionBroadcast  [ 0 ]
  22: CPhysicalMotionHashDistribute HASHED: [ CScalarIdent "a" (36), nulls colocated ], opfamilies: (1977,1.0), [ 0 ]
```

**Comment by RekGRpth** [comment]
> > When translating expression to DXL, ORCA sets one segment for input array if strict or tainted replicated
> 
> This must be described in the PR description in text form. Links and code snippets are only useful addition.

done

**Comment by Stolb27** [comment]
>> Please, prove it.

> comparing Physical plan before PR and Physical plan after PR shows that CPhysicalMotionBroadcast from group 0 expression 21 changed to CPhysicalMotionHashDistribute from group 0 expression 22 while group 0 is same before and after PR.

All `CPhysicalInnerHashJoin` operators refer to another groups. So, discussed Redistribute Motion from physical plan isn't connected with this group. Members of this group were added by another enforcer. I'm suspect that resulting physical plan is another alternative. So the question is still here, how was your enforcer is applied?

**Comment by RekGRpth** [comment]
> All `CPhysicalInnerHashJoin` operators refer to another groups

these groups are same also:
```sql
Group 5 (#GExprs: 5):
  0: CLogicalNAryJoin [ 0 1 4 ]
  1: CLogicalInnerJoin [ 0 1 4 ]
  2: CLogicalInnerJoin [ 1 0 4 ]
  3: CPhysicalInnerHashJoin (High) [ 1 0 4 ]
  4: CPhysicalInnerHashJoin (High) [ 0 1 4 ]

Group 18 (#GExprs: 10):
  0: CLogicalInnerJoin [ 8 9 13 ]
  1: CLogicalInnerJoin [ 9 8 13 ]
  2: CPhysicalInnerHashJoin (High) [ 9 8 13 ]
  3: CPhysicalInnerHashJoin (High) [ 8 9 13 ]
  4: CPhysicalMotionGather(master) [ 18 ]
  5: CPhysicalMotionRandom [ 18 ]
  6: CPhysicalMotionHashDistribute HASHED: [ CScalarIdent "b" (37), nulls colocated ], opfamilies: (1977,1.0), [ 18 ]
  7: CPhysicalMotionBroadcast  [ 18 ]
  8: CPhysicalMotionHashDistribute HASHED: [ +--CScalarIdent "b" (37)   origin: [Grp:11, GrpExpr:0]
                                        +--CScalarIdent "b" (45)   origin: [Grp:12, GrpExpr:0]
                                        , nulls colocated ], opfamilies: (1977,1.0),(1977,1.0), [ 18 ]
  9: CPhysicalMotionHashDistribute HASHED: [ CScalarIdent "b" (45), nulls colocated ], opfamilies: (1977,1.0), [ 18 ]
```

**Comment by RekGRpth** [comment]
> discussed Redistribute Motion from physical plan

I mean 

> ```sql
> CPhysicalMotionBroadcast    rows:1   width:46  rebinds:1   cost:1293.000786   origin: [Grp:0, GrpExpr:21]
> ```

and

> ```sql
> CPhysicalMotionHashDistribute HASHED: [ CScalarIdent "a" (36), nulls colocated ], opfamilies: (1977,1.0),   rows:1   width:46  rebinds:1   cost:1293.000720   origin: [Grp:0, GrpExpr:22]
> ```

**Comment by RekGRpth** [comment]
> resulting physical plan is another alternative

yes, without broadcast expression 21 from group 0 and with redistribute expression 22 from group 0

**Comment by RekGRpth** [comment]
> how was your enforcer is applied?

none enforcer is applied, patch disables broadcast enforcer for some conditions

**Pull request closed without merging**